docs/snapshotting/design.md

```diff
@@ -0,0 +1,217 @@
+- guest's storage disks' contents.
+
+The current snapshot proposal aims for increased flexibility by independently generating and providing the state of components listed above. This means that taking a snapshot will result in multiple files that are composing the full microVM snapshot:
+- guest memory file,
+- vmm state file,
+- zero or more disk files (depending on how many the guest had).
+
+This design allows sharing any read-only resources (such as read-only disks or vmm state) between multiple microVMs to improve density.
+When loading a snapshot, instead of loading the full contents from file to memory at resume time, Firecracker will create a `MAP_PRIVATE` mapping of the memory file resulting in runtime on-demand loading of memory pages. Any subsequent memory writes go to a copy-on-write anonymous memory mapping.
+
+This has the advantage of **very fast snapshot loading** times, but comes with the cost of having to keep the _guest memory file_ around for the entire lifetime of the resumed microVM.
+
+## Key Design Choices
+
+- Fresh Firecracker microVMs will be booted using `anonymous` memory, while microVMs resumed from snapshot will load memory on-demand from the snapshot and copy-on-write to anonymous memory.
+- *Resuming from a snapshot* is optimized for speed. *Taking a snapshot* will always involve synchronously writing dirty memory pages to snapshot file. Taking a snapshot of a fresh microVM running off non-file-backed anonymous memory, will result in full contents of guest memory being written to the snapshot.
+- The snapshot operations have minimal requirements/dependencies on when/how they're used.
+- The API calls exposing the snapshotting functionality have clear Prerequisites that describe the previously mentioned minimal requirements. Any flow that doesn't violate these prerequisites should be allowed and supported.
+- _Memory file_ and _Vmm state file_ are generated by Firecracker on snapshot creation and the disk contents are flushed to their backing files. The disk files will be **managed by the user**; users need to externally back up their block devices backing files.
+
+## API Definition
+An extra flag is added to the `/machine-config` resource, to specify whether the microVM will support incremental snapshots or not. Enabling this support will enable KVM dirty page tracking so it comes at a cost. Only use it when needed.
+
+Supported individual operations are: `Pause`, `Resume`, `CreateFullSnapshot`, `CreateDiffSnapshot` and `LoadFullSnapshot`. To create a snapshot, pause the vm; you can resume it afterwards. Full snapshots always create a complete, resume-able snapshot of the current microVM state and memory. Diff snapshots save the current microVM state and the memory dirtied since the last snapshot (full or diff). Diff snapshots are not resume-able, but can be merged into a full snapshot using external (provided) tooling. The HTTP APIs are:
+
+### `PUT /pause`: will attempt to pause the microVM
```

Comment:
This doesn't seem very RESTful. Why is this not a `PATCH` on `/vm` or something like that? It doesn't seem like a separate object.

Reply:
We represented `/vm` as `/` since all api urls refer to this vm.

If actions/state-changes don't look RESTful in the URL, we can move `pause` and `resume` next to the `InstanceStart` command since these 3 belong together (currently under `/actions`; wherever that may be long-term). https://github.com/firecracker-microvm/firecracker/blob/59584b18faa55f6fe4c4aa5418cf9904652e5d3d/src/api_server/swagger/firecracker.yaml#L502

We have a goal to redesign the whole API. In the meantime I would like to at least keep it consistent.

Reply:
I would prefer to not continue adding other options to `/actions` because in my opinion `/actions` shouldn't be a resource anyway. I like the `PATCH` option on a `state` field of `vm` and I think that moving the `/actions` variants will be cleaner in the future with this approach (if, for example, an action can't be mapped to the `state` field, we can add another field that would make more sense and call `PATCH` on that field; I suppose this would be the `FlushMetrics` case).
There is also the `PUT /vm/some-action` option that won't contain a request body, but this solution would involve a lot of new operations in our `swagger` specification that won't really belong to a defined `vm` object (even if we have the `vm` word in the path).

Reply:
It is true this is not RESTful and I would call it a RPC/RESTful hybrid API. 
Looking at the larger picture here from an user experience point of view, the interaction with the VMM is all action based while only our internal implementation is designed around resources. As a user I would use this flow to start a FC microvm :
- start firecracker vmm.
- configure cpu and memory
- attach a block devices
- attach a net device
- start microvm.

To me, this looks like RPC and not REST.

That being said,  what is the better option here in terms of implementation and consistency with respect to our API redesign plan in the future ?

Reply:
For what it's worth, I like `PUT /vm/<some-action>` the most even if it's not RESTful. It's really easy to use and to understand, easier to identify/debug in logs, etc.
It also more extensible and allows adding `a-new-action` which needs data in its body.

Reply:
I understand that it is easier to use because you won't have a body for that request but it is also pretty unusual for a `PUT` request to not have a body and to me such request wouldn't make much sense.

Reply:
>it is also pretty unusual for a PUT request to not have a body and to me such request wouldn't make much sense

And you're not the only one; some software do it, some say they shouldn't: https://stackoverflow.com/questions/1233372/is-an-http-put-request-required-to-include-a-body

Either way, I am subscribing to @sandreim's comment:
> That being said, what is the better option here in terms of implementation cost and consistency with respect to our API redesign plan in the future ?

For now, we can choose **any one** of the current options as long as we offer a consistent API.

Reply:
I agree our API shouldn't be RESTful and in the end, for the particular `pause` and `resume` problem, both solutions are more or less workarounds. Anyway, in both cases, to offer a consistent API, `/actions` should be removed, so from this point of view I don't see an option to be better than the other.

Reply:
I updated #1821 with the `PATCH /vm` option. Please take a look!